src/groupeddataframe/grouping.jl

```diff
@@ -122,50 +119,111 @@ _names(gd::GroupedDataFrame) = _names(gd.parent)
 # Wrapping automatically adds column names when the value returned
 # by the user-provided function lacks them
 wrap(x::Union{AbstractDataFrame, NamedTuple, DataFrameRow}) = x
-wrap(x::AbstractMatrix) = convert(DataFrame, x)
-wrap(x::AbstractVector) = DataFrame(x1 = x)
-wrap(x::Any) = (x1 = x,)
+wrap(x::AbstractMatrix) =
+    NamedTuple{Tuple(gennames(size(x, 2)))}(Tuple(x[:, i] for i in 1:size(x, 2)))
+wrap(x::Any) = (x1=x,)
+
+function wrap_call(f::Union{Function, Type}, gd::GroupedDataFrame,
+                   incols::Tuple, i::Integer)
+    idx = gd.idx[gd.starts[i]:gd.ends[i]]
+    wrap(f(map(c -> view(c, idx), incols)...))
+end
+function wrap_call(f::Union{Function, Type}, gd::GroupedDataFrame,
+                   incols::NamedTuple, i::Integer)
+    idx = gd.idx[gd.starts[i]:gd.ends[i]]
+    wrap(f(map(c -> view(c, idx), incols)))
+end
+wrap_call(f::Union{Function, Type}, gd::GroupedDataFrame, incols::Nothing, i::Integer) =
+    wrap(f(gd[i]))
+
+_nrow(df::AbstractDataFrame) = nrow(df)
+_nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
+    isempty(x) ? 0 : length(x[1])
+_ncol(df::AbstractDataFrame) = ncol(df)
+_ncol(x::Union{NamedTuple, DataFrameRow}) = length(x)
 
 """
-    map(f::Union{Function, Type}, gd::GroupedDataFrame)
+    map(cols => f, gd::GroupedDataFrame)
+    map(f, gd::GroupedDataFrame)
 
 Apply a function to each group of rows and return a `GroupedDataFrame`.
 
-For each group in `gd`, `f` is passed a `SubDataFrame` view holding the corresponding rows.
+If the first argument is a `cols => f` pair, `cols` must be a column name or index, or
+a vector or tuple thereof, and `f` must be a function. If `cols` is a single column index,
```

Comment:
A corner case: can `cols` be a vector of `Bool` to indicate columns?

Reply:
See other thread. I'd rather not mention this explicitly here since that's relatively unusual and it would make the sentence really complex. It would be interesting to define and export a type alias and use it in signatures, like `ColumnIndex` but including vectors (JuliaDB uses `::Selection`, but doesn't export it).

Reply:
OK - but let us at least internally be clear if we want to support `AbstractVector{Bool}` (see other comment where the implementation is provided).